### ISSUE TITLE
Clarity on the Snapshot behaviour

### ISSUE THREAD

**Issue by Vivek205** [snapshot-labs/snapshot-strategies#209]
I have a query on how the snapshot uses blockTag in the option. 
This issue is a continuation of the discussion below
https://github.com/snapshot-labs/snapshot-strategies/pull/202#issuecomment-982582827
---
@ChaituVR 
We have custom staking contract and we decide the users' voting power by calling a get method in the contract.
How will snapshot.js will retrieve the past data based on the blocktag for this scenario?

Are you storing the data for the users anywhere in a DB?

reg:

ex: if I vote with 1 VP and in future, I may buy 100 VP and change the results)

if the user has already voted with 1VP, won't it be persisted in the DB and won't the user loses his ability to vote for the proposal thereafter?
Even if the user buys more tokens after voting, it should not impact the proposal because he has already cast his vote.

**Comment by ChaituVR** [comment]
Hello @Vivek205 I said this when you are passing latest into the blocktag

This could cause issues while calculating the scores, they may buy VP based on the proposal and change the results if so. And snapshot doesn't always store voting power, it will only store only once after the proposal is ended, and there are methods like `getScores` in snapshot.js which calculates voting power (which doesn't use stored voting power) 

So i suggested to pass block number into the strategy, which will help strategy to return the voting power consistently based on the block number.

**Comment by Vivek205** [comment]
@ChaituVR 
Does that mean, the user will be able to vote multiple times in a proposal? 
And every time he votes, will his existing vote be overwritten?

**Comment by raamb** [comment]
Hi @ChaituVR just to be clear, would like to illustrate this further.

Monday : Proposal 1 Opens
Tuesday: User u1 own 10 tokens and votes on Proposal 1
Wednesday: User u1 buys additional 10 tokens (total 20 tokens)
Thursday: Proposal 1 closes

When the votes are tallied, do we count user u1's vote as 10 or 20?

**Comment by ChaituVR** [comment]
Hello @Vivek205 @raamb 
> Does that mean, the user will be able to vote multiple times in a proposal?
And every time he votes, will his existing vote be overwritten?

- A voter can change their vote (selected choices) any time within the voting period, but voting power will not be changed - voting power is based on the strategies and block number of the proposal - it is always the same, and once the voting is closed user cannot change the choice
> Monday: Proposal 1 Opens
Tuesday: User u1 own 10 tokens and votes on Proposal 1
Wednesday: User u1 buys additional 10 tokens (total 20 tokens)
Thursday: Proposal 1 closes 
When the votes are tallied, do we count user u1's vote as 10 or 20?

- u1 voting power will be 10 even after the voting is closed

**Comment by raamb** [comment]
Thanks @ChaituVR what you mention makes sense when
- Voting power is based only on account balance
- Snapshot balance is set to a past block-number

In our case, voting power is not just a function of account balance but also of staking balance and Liquidity provisioning. 
For staking balance and Liquidity provisioning there is no obvious way to get the balance as of some past block number unless somebody explicitly persists the values (and we don't do that). 
In this scenario for a user 
> voting power = balance of user as of a past block-number + staking balance as of when the proposal ends + Liquidity provisioning as of when the proposal ends.

This becomes complicated and non-intuitive. 
Does my example make sense? How do others handle staking and liquidity provisioning. Thanks a lot for your help with these questions of ours

**Comment by ChaituVR** [comment]
> For staking balance and Liquidity provisioning there is no obvious way to get the balance as of some past block number 

Why not? 🤔 
All your three strategies 
```
snet-farmers
snet-liquidity-providers
snet-stakers
```
all are able to fetch balances/staking balance/liquidity from a particular block right? 

> voting power = balance of user as of a past block-number + staking balance as of when the proposal ends + Liquidity provisioning as of when the proposal ends.

This may not be possible as there is no way to pass block numbers when proposal ends
But we can do
```
voting power = 
balance of user at the proposal block 
+ staking balance at the proposal block 
+ Liquidity provisioning at the proposal block 
```

**Comment by Vivek205** [comment]
@ChaituVR 
The staking and farming balances are retrieved from the state of the contract. 
These contracts won't return the data for a past block number. 
They always return the data at the latest instance of the blockchain. 

So, the staking and the farming values captured during the start of the proposal might not be relevant when the voting proposal ends.

**Comment by ChaituVR** [comment]
> These contracts won't return the data for a past block number.

Snapshot read the state of the contract at the past block number

**Comment by ChaituVR** [comment]
@Vivek205 @raamb For better understanding try a test proposal with this strategies :)

**Comment by ChaituVR** [comment]
Closing this for now, please feel free to reopen or create new issue @Vivek205
